r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -227,10 +360,30 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
             )?;
         }
 
+        // Assumes that 
+        // ``
+        //     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+        // ``
+        // where `surfeit' = log(num_limbs + num_adds + 1)`.
+        // Costs
+        // ``
+        //      (num_groups - 1) * (1 + 2*bits_per_limb + surfeit' + 2 - bits_per_limb) + 2
+        // ``
+        // constraints, where `num_groups = Ceil[num_limbs / S]`, and
+        // ``
+        //  S - 1 = Floor[
+        //          (ConstraintF::CAPACITY - 2 - (2 * bits_per_limb + surfeit') / bits_per_limb
+        //      ] 
+        //        = Floor[
+        //          (ConstraintF::CAPACITY - 2 - surfeit') / bits_per_limb
+        //      ] - 2.
+        // ``
         Reducer::<SimulationF, ConstraintF>::group_and_check_equality(
             cs.ns(|| "group and check equality"),
-            surfeit,
+            surfeit_kp_plus_r,
```

Comment:
Why does this depend only on the surfeit of k * p + r?

Reply:
Because `num_adds + num_limbs + 1` is bigger than `num_adds`.